web-ui/src/streaming/builder/hooks/useAutoLayout.ts

```diff
@@ -0,0 +1,229 @@
+  let sortedIndex = 0
+  for (let i = 0; i < nodes.length; i++) {
+    if (!nodesAndIndexWithoutAnEdge[i]) {
+      result.push(nodesWithAnEdge[sortedIndex])
+      sortedIndex++
+    } else {
+      result.push(nodes[i])
+    }
+  }
+
+  return result
+}
+
+// Positions the nodes in a fixed layout. The layout is:
+//
+// - Input nodes are on the left
+// - Output nodes are on the right
+// - Program node is in the middle
+// - The Input and Output nodes are evenly spaced vertically.
+// - The program node is centered vertically wrg. to the input and output nodes.
+// - Input nodes are reordered based on which table they are connected to.
+//
+// The input and output nodes are evenly spaced horizontally from the program
+// node. The output nodes are left aligned, the input nodes are right-aligned.
+function layoutNodesFixed(
+  getNodes: Instance.GetNodes<any>,
+  setNodes: Instance.SetNodes<any>,
+  getEdges: Instance.GetEdges<any>
+) {
+  const nodes: Node[] = getNodes()
+  const programNode = nodes.filter(node => node.type === 'sqlPlaceholder' || node.type === 'sqlProgram')
+  assert(programNode.length === 1)
+
+  const rawInputNodes = nodes.filter(node => node.type === 'inputNode')
+  const inputOrder = new Map<string, number>()
+  if (programNode[0].data?.program?.schema) {
+    programNode[0].data.program.schema.inputs.forEach((element: { name: string }, index: number) => {
+      inputOrder.set(element.name, index)
+    })
+  }
+  const inputNodes = sortInputConnectors(rawInputNodes, getEdges(), inputOrder).concat(
+    nodes.filter(node => node.id === 'inputPlaceholder')
+  )
+
+  const outputNodes = nodes
+    .filter(node => node.type === 'outputNode' || node.id === 'outputPlaceholder')
+    .sort((a, b) => a.position.y - b.position.y)
+
+  // Vertical adjustment
+  //
+  // Calculate how much vertical space input/output nodes consume (not
+  // including the placeholder nodes at the end).
+  //
+  // We want to:
+  // - Keep the input and output nodes evenly spaced.
+  // - Keep the program window centered vertically wrg. to the input and output nodes
+  const totalInputHeight = inputNodes
+    .filter(node => node.type === 'inputNode')
+    .map(node => {
+      return (node.height || 0) + VERTICAL_SPACING
+    })
+    .reduce((a, b) => a + b, 0)
+  const totalOutputHeight = outputNodes
+    .filter(node => node.type === 'outputNode')
+    .map(node => {
+      return (node.height || 0) + VERTICAL_SPACING
+    })
+    .reduce((a, b) => a + b, 0)
+
+  // Horizontal adjustment
+  //
+  // We right align the input column based on the widest input node. We make
+  // spavce between input/program and program/output is the same.
+  const maxInputWidth = inputNodes.map(node => node.width || 0).reduce((a, b) => Math.max(a, b), 0)
+  const programWidth = programNode.map(node => node.width || 0).reduce((a, b) => Math.max(a, b), 0)
+  const programHeight = programNode.map(node => node.height || 0).reduce((a, b) => Math.max(a, b), 0)
```

Comment:
do these nodes have a maximum width?
They should not be too wide, e.g., if view names are long.

Reply:
It should layout correctly for arbitrary long names, the user can zoom in-out using the mouse wheel if space is an issue.